src/shims/sync.rs

```diff
@@ -0,0 +1,438 @@
+            let attr_place = this.deref_operand(attr_op)?;
+            let i32_layout = this.layout_of(this.tcx.types.i32)?;
+            let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+            this.write_scalar(kind, kind_place.into())?;
+        } else {
+            let einval = this.eval_libc_i32("EINVAL")?;
+            return Ok(einval);
+        }
+
+        Ok(0)
+    }
+
+    fn pthread_mutexattr_destroy(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        if this.is_null(attr)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let attr_place = this.deref_operand(attr_op)?;
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+        let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+        this.write_scalar(ScalarMaybeUndef::Undef, kind_place.into())?;
+
+        Ok(0)
+    }
+
+    // pthread_mutex_t is between 24 and 48 bytes, depending on the platform
+    // memory layout:
+    // bytes 0-3: reserved for signature on macOS
+    // bytes 4-7: count of how many times this mutex has been locked, as a u32
+    // bytes 12-15: mutex kind, as an i32
+    // (the kind should be at this offset for compatibility with the static
+    // initializer macro)
+
+    fn pthread_mutex_init(
+        &mut self,
+        mutex_op: OpTy<'tcx, Tag>,
+        attr_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        check_ptr_target_min_size(this, mutex_op, 16)?;
+        check_ptr_target_min_size(this, attr_op, 4)?;
+
+        let mutex = this.read_scalar(mutex_op)?.not_undef()?;
+        if this.is_null(mutex)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+        let mutex_place = this.deref_operand(mutex_op)?;
+
+        let i32_layout = this.layout_of(this.tcx.types.i32)?;
+
+        let attr = this.read_scalar(attr_op)?.not_undef()?;
+        let kind = if this.is_null(attr)? {
+            this.eval_libc("PTHREAD_MUTEX_DEFAULT")?
+        } else {
+            let attr_place = this.deref_operand(attr_op)?;
+            let attr_kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, i32_layout, this)?;
+            this.read_scalar(attr_kind_place.into())?.not_undef()?
+        };
+
+        let u32_layout = this.layout_of(this.tcx.types.u32)?;
+        let locked_count_place = mutex_place.offset(Size::from_bytes(4), MemPlaceMeta::None, u32_layout, this)?;
+        this.write_scalar(Scalar::from_u32(0), locked_count_place.into())?;
+
+        let mutex_kind_place = mutex_place.offset(Size::from_bytes(12), MemPlaceMeta::None, i32_layout, &*this.tcx)?;
+        this.write_scalar(kind, mutex_kind_place.into())?;
```

Comment:
So you basically have some form of data structure here, ignoring the actual `pthread_mutex_t` of the current platform, with two fields at offsets 4 and 12...
To make this clearer, could you have some private helper methods as setters/getters for these fields? (Same for the pthread_mutexattr_t above.) Ideally, these "high-level" methods here that implement the shims, shouldn't have to be concerned at all with these annoying data layout details.